src/store/api/worldUsers.ts

```diff
@@ -0,0 +1,231 @@
+            "\n  snapshot.docChanges() modified:",
+            modified,
+            "\n  queuedChanges.length",
+            queuedChanges.length
+          );
+
+          snapshot.docChanges().forEach((change) => {
+            // TODO: check if this document relates to the current user, if so, update it immediately
+            //   Note that we will need to pass their userId in and/or read it from somehwere to make this work..
+            //   probably just the queryArgs for this whole api listener
+            //
+            // const currentUserId = "TODO";
+            //
+            // if (change.doc.id === currentUserId) {
+            //   updateCachedData((draft) => {
+            //     processUserDocChange(draft)(change);
+            //   });
+            // } else {
+            queuedChanges.push(change);
+            // }
+          });
+        });
+
+        // Wait until the data no longer needs to be kept in our redux cache
+        await cacheEntryRemoved;
+
+        // Unsubscribe the firestore query snapshot listener
+        unsubscribeListener();
+
+        // Clear the interval for processing the queued changes
+        clearInterval(processQueuedChangesIntervalId);
+
+        // Make sure we process any last remaining queued changes
+        processQueuedChanges();
+      },
+    }),
+  }),
+});
+
+// TODO: https://redux-toolkit.js.org/rtk-query/api/created-api/hooks
+// TODO: https://redux-toolkit.js.org/rtk-query/api/created-api/hooks#skiptoken
+export const {
+  useQuery: useWorldUsersQuery,
+  useQueryState: useWorldUsersQueryState,
+  useQuerySubscription: useWorldUsersQuerySubscription,
+} = worldUsersApi.endpoints.worldUsers;
+
+const processUserDocChange = (draft: MaybeDrafted<WorldUsersData>) => (
+  change: firebase.firestore.DocumentChange<firebase.firestore.DocumentData>
+) => {
+  // @debt validate/typecast this properly so we don't have to use 'as' to hack the types here
+  const user: UserWithLocation = change.doc.data() as UserWithLocation;
+  const userId: string = change.doc.id;
+  const userWithId: WithId<UserWithLocation> = withId(user, userId);
+
+  const userWithoutLocation: WithId<User> = userWithLocationToUser(userWithId);
+
+  const userLocation: WithId<UserLocation> = extractLocationFromUser(
+    userWithId
+  );
+
+  switch (change.type) {
+    case "added": {
+      // TODO: theoretically I believe it should never be possible for a duplicate user to end up here from
+      //   this, but I wonder if we should err on the side of caution and combine the added/modified cases to
+      //   always try and find the existing user first? A little extra overhead for potentially a little
+      //   more safety.
+      draft.worldUsers.push(userWithoutLocation);
+      draft.worldUsersById[userId] = userWithoutLocation;
+      draft.worldUserLocationsById[userId] = userLocation;
+
+      return;
+    }
+
+    case "modified": {
+      const existingUserIndex = draft.worldUsers.findIndex(
```

Comment:
[note] `processUserDocChange` handles the actual logic for making updates to the datastore based on the 'change' data received previously. This makes use of Immer's `draft`, that is created earlier with `updateCachedData`.